sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/GpuHashJoin.scala

```diff
@@ -511,6 +507,208 @@ class ConditionalHashJoinIterator(
+        val gatherer = JoinGatherer(lazyLeftMap, leftData, lazyRightMap, rightData,
+          OutOfBoundsPolicy.DONT_CHECK, OutOfBoundsPolicy.NULLIFY)
+        if (gatherer.isDone) {
+          // Nothing matched...
+          gatherer.close()
+          None
+        } else {
+          updateTrackingMask(lazyRightMap)
+          Some(gatherer)
+        }
+      } finally {
+        maps.foreach(_.close())
+      }
+    }
+  }
+
+  override def close(): Unit = {
+    if (!closed) {
+      super.close()
+      compiledConditionRes.foreach(_.close())
+      rightTracker.foreach(_.close())
+    }
+  }
+
+  override def hasFinalBatch(): Option[ColumnarBatch] = {
+    rightTracker match {
+      case None => None
+      case Some(tracker) => {
+        val filteredBatch = withResource(tracker.releaseBatch()) { trackerBatch =>
+          withResource(GpuColumnVector.from(trackerBatch)) { trackerTab =>
+            val batch = built.getBatch
+            withResource(GpuColumnVector.from(batch)) { builtTable =>
+              withResource(builtTable.filter(trackerTab.getColumn(0))) { filterTab =>
+                GpuColumnVector.from(filterTab, GpuColumnVector.extractTypes(batch))
+              }
+            }
+          }
+        }
+        // Combine right columns with null columns for left side
+        val ret = withResource(filteredBatch) { rightBatch =>
+          val numFilterRows = rightBatch.numRows()
+          if (numFilterRows > 0) {
+            val leftColumns = streamAttributes.map { attr =>
+              GpuColumnVector.fromNull(numFilterRows, attr.dataType)
+            }
+            withResource(new ColumnarBatch(leftColumns.toArray, numFilterRows)) { leftBatch =>
+              Some(GpuColumnVector.combineColumns(leftBatch, rightBatch))
+            }
+          } else {
+            None
+          }
+        }
+        rightTracker = None
+        ret
+      }
+    }
+  }
+
+  private def trueColumnTable(numRows: Int): Table = {
+    withResource(Scalar.fromBool(true)) { trueScalar =>
+      withResource(ai.rapids.cudf.ColumnVector.fromScalar(trueScalar, numRows)) {
+        new Table(_)
+      }
+    }
+  }
+
+  // Create a boolean column to indicate which gather map rows are valid.
+  private def validIndexMask(gatherView: ColumnView): ColumnView = {
+    withResource(Scalar.fromInt(-numRightRows)) { minValue =>
+      withResource(Scalar.fromInt(numRightRows)) { maxValue =>
+        withResource(gatherView.greaterOrEqualTo(minValue)) { greaterEqualMin =>
+          withResource(gatherView.lessThan(maxValue)) { lessMax =>
+            lessMax.and(greaterEqualMin)
+          }
+        }
+      }
+    }
+  }
+
+  /**
+   * Update the tracking mask.
+   */
+  private def updateTrackingMask(rightGatherMap: LazySpillableGatherMap): Unit = {
+    // Get the current tracking table, or all true table to start with
+    val rightTrackingTable = rightTracker.map { spillableBatch =>
```

Comment:
This leaks if something throws before we get to the `withResource(rightTrackingTable)` below.  Seems like this is being created too early.

Reply:
Fixed.